dashboard/app/views/levels/_summary.html.haml

```diff
@@ -11,6 +11,7 @@
     {
       user_id: response.user_id,
       text: response.level_source&.data,
+      student_name: User.find_by(id: response.user_id)&.name,
```

Comment:
I am looking for extra scrutiny here.  I didn't see anywhere else in our codebase where this approach was used.  BUT I had a really hard time finding a better way to get this data in here.  I also considered matching up student names and ids in the react component (essentially saying "hey, you have a list of `student names` in this section, match them to the `student ids` connected to the `responses`"

I really wanted to get this data in earlier though.  I'd love to hear how we feel about this approach or any other way that might be more aligned to best practices.

Reply:
This is a very good first step. I think there is a slightly better approach that doesn't require the extra lookup to find the user based on the ID.

The @responses here is a list of UserLevels ([see where it's set](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/script_levels_controller.rb#L177))

A [user level](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/user_level.rb#L35) has a :user on it

I think we should then be able to do `response.user.name` (similar to [here](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/peer_review.rb#L268))

The only reason this is better is because we don't need to do an extra `User.find_by` query. Although I am by no means a ruby expert, so if you want to check me or want more context, I'd ask Eric

Reply:
Thanks for the written explanation!  I do want to chat a bit more about how you found this tomorrow in our 1-1.  I was looking for something like this (... for probably way too long...) but despite my sleuthing this didn't show up on my radar.  Thanks for the digging - this better.

Reply:
By default Rails doesn't know there's anything special about this `response.user_id` compared to any other numeric field. A valid way to look up the user would then be exactly what you did originally: `User.find_by(id: response.user_id)` (or its shorter cousin `User.find(response.user_id)`).

However this `find` syntax is a bit cumbersome. Rails gives a way to define [associations](https://guides.rubyonrails.org/association_basics.html) between models. These let you declare the relationships between tables in your database schema and translate them into properties you can call on the model objects (`response.user` in this case).

It's generally a good bet that when a database table has a `user_id` someone has probably already set up the necessary association linking the models together in this way. Might as well try using it!

Liam already pointed to the code where this happens in this case. I noticed something a bit unusual there: the `optional: true` at the end of the association declaration. This is saying that the `User` query might not actually return anything, and that Rails should return `nil` instead of raising an error in this case. Probably to handle soft-deleted users? Best to use `response.user&.name` to avoid some Honeybadger errors (and also handle this case on the front end so the UI says something like "name not found" instead of "written by nil" or whatever).

In terms of queries, Rails has "lazy loading" which means that when you load a `UserLevel` it won't query for the associated `User` or other associated objects until you actually ask for some field from that object.

Look at this Rails console output:
<img width="794" alt="image" src="https://github.com/user-attachments/assets/bacf85d5-c88c-415f-9fd1-686790971557">

We can see a few things here. First, Rails doesn't query the `users` table when I first load the `UserLevel`...instead it waits until I call `response.user.name`. Second, when I call `response.user.family_name` it doesn't query the database again. The results from the prior query for that user are still stored in memory. Third, Rails issues the exact same query for `response.user.name` as it does for `User.find_by(id: response.user_id).name`. So, going through the Rails association might save you a query if some other part of the code already is looking at something in `response.user` during this web request. Even if not, you're no worse off.

Fourthly, I also made a mistake in my `find_by` query, passing in `response.id` instead of `response.user_id`. Going through the defined association instead of the lower-level query can save bugs too. :-)

Reply:
Thanks so much for taking time to write this up, Eric.  This was so helpful in filling in the blanks!